src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,127 @@
+        size = 0;
+    }
+
+    public void growIfArrayIsFull() {
+        if (size == elements.length) {
+            Object[] newArray = new Object[Math.round(elements.length * 1.5f)];
+            for (int i = 0; i < elements.length; i++) {
+                newArray[i] = elements[i];
+            }
+            elements = newArray;
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayIsFull();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index didn't exist");
+        }
+        if (elements[index] == null) {
+            add(value);
+        } else {
+            growIfArrayIsFull();
+            size++;
+            for (int i = size - 1; i > index; i--) {
+                elements[i] = elements[i - 1];
+            }
+            elements[index] = value;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index didn't exist");
+        } else {
+            return (T) elements[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index didn't exist");
+        } else {
+            elements[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index didn't exist");
+        } else {
+            final Object element = elements[index];
+            for (int i = index; i < size - 1; i++) {
+                elements[i] = elements[i + 1];
+            }
+            elements[size - 1] = null;
+            size--;
+            return (T) element;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < size; i++) {
+                if (elements[i] == null) {
+                    for (int j = i; j < size; j++) {
+                        elements[j] = elements[j + 1];
+                    }
+                    size--;
+                    return element;
+                }
+            }
+        }
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null && elements[i].equals(element)) {
+                for (int j = i; j < size; j++) {
+                    elements[j] = elements[j + 1];
+                }
+                size--;
+                return element;
+            }
+        }
+        throw new NoSuchElementException();
```

Comment:
The NoSuchElementException should have an informative message such as 'Element not found: ' + element.